# Wrangle and Analyze Data Project

## Introduction

The purpose of this project is to wrangle data from WeRateDogs Twitter account. First, data is going to be obtained
from multiple sources. Next, it is going to be assessed in terms of both [quality and tidiness](https://ryanwingate.com/purpose/tidy-data/ ). Finally, I will clean the data so that it is ready for the analysis.

## Gathering Data

In [1]:
# import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import requests
import os

In [2]:
# load Twitter archive data and display 1 row
df_archive_original = pd.read_csv('twitter-archive-enhanced.csv')
df_archive_original.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [3]:
# Download the tweet image predictions
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
local_filename = url.split('/')[-1]
r = requests.get(url)
f = open(local_filename, 'wb')
f.write(r.content)

335079

In [4]:
# open and look at the predictions
df_pred_original = pd.read_csv('image-predictions.tsv', sep='\t')
df_pred_original.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
# Check for tweet_json.txt
# if tweet_json.txt doesn't exist, download programatically
url = 'https://s3.amazonaws.com/video.udacity-data.com/topher/2018/November/5be5fb7d_tweet-json/tweet-json.txt'
json_txt_file_path = 'tweet_json.txt'
if not os.path.isfile(json_txt_file_path):
    with open(json_txt_file_path, mode = 'wb') as file:
        file.write(requests.get(url).content)
# Load into DataFrame
df_json_original = pd.read_json(json_txt_file_path, lines=True)
df_json_original = df_json_original[['id', 'retweet_count', 'favorite_count']]
df_json_original.head()

id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

Make copies of the dataframes for the analysis.

In [6]:
df_archive = df_archive_original.copy()
df_pred = df_pred_original.copy()
df_json = df_json_original.copy()

## Assessing Data

In [7]:
df_archive.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [8]:
df_pred.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [9]:
df_json.head(1)

id  retweet_count  favorite_count
0  892420643555336193           8853           39467

In [10]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [11]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [12]:
df_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
id                2354 non-null int64
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


In [13]:
df_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [14]:
df_archive.name.value_counts()

None         745
a             55
Charlie       12
Lucy          11
Oliver        11
Cooper        11
Penny         10
Tucker        10
Lola          10
Bo             9
Winston        9
the            8
Sadie          8
Buddy          7
Toby           7
Bailey         7
Daisy          7
an             7
Leo            6
Scout          6
Jax            6
Stanley        6
Jack           6
Rusty          6
Koda           6
Milo           6
Dave           6
Oscar          6
Bella          6
George         5
            ... 
Karma          1
Sprinkles      1
Caryl          1
Lupe           1
Kallie         1
Huck           1
Tom            1
Andy           1
Saydee         1
Sully          1
Cheesy         1
DayZ           1
Tess           1
Alexander      1
Coopson        1
Amélie         1
Strudel        1
Lizzie         1
Cilantro       1
Tommy          1
Timofy         1
Autumn         1
Pavlov         1
Gerbald        1
Snickers       1
Fletcher       1
Ebby           1
Robin         

### Quality

1. Change the **id** column in `df_json` dataframe to **tweet_id** in order to match it with other dataframes. This will make is easier to merge dataframes at the last steps. In addition, I will change the data type from int to string because ids should not be numeric and they aren't intended to perform calculations.
2. Change **timestamp** data type in `df_archive`. This step will help if I was to analyze the data based on time or date.
3. Drop **retweets** rows in `df_archive`. Only original tweets must be included in the final master dataset.
4. Drop unnecessary columns in `df_archive`. It is not properly an issue, but it will make it easier to analyze the data.
5. Drop null values from **expanded_urls** in `df_archive`. The **expanded_urls** is the only column that has a few missing values. I could not determine why that was the case nor I could fill these values out; therefore, I will drop these values.
6. Drop tweets that are not dogs. This analysis should only incorporate tweets of dogs; therefore, the next step is to delete rows that contain non-dog tweets. In order to do that, I will merge the df_pred dataframe with df_archive and create a new dataframe df_archive2. The df_pred contains 3 predictions from the neural network whether a picture shows a dog. I assume that if any of the predictions suggest that it is a dog, I treated it as a dog. I checked a few tweets ‘by hand’ and it appeared to be true.
7. Clean **rating_denominator** in `df_archive`. Extract the values from text, delete tweets without ratings and tweets with multiple dogs. Change the data type to float.
8. Clean **rating_nominator** in `df_archive`. Extract the values from text, delete tweets without ratings and tweets with multiple dogs. Change the data type to float.
9. Clean names using regular expression in df_archive. A useful pattern to find the incorrect names is that they started with lowercase, instead, the true names always started with uppercase. In addition, the real names usually are placed after words like "named", "is" etc.


### Tidiness
1. Melt 'doggo' 'floofer' 'pupper' 'puppo' in `df_archive`.
2. Join all data frames into `twitter_archive_master.csv`.

## Cleaning Data
### Quality

**Define**
1. Change the **id** column in `df_json` dataframe to **tweet_id** in order to match it with other dataframes. This will make is easier to merge dataframes at the last steps. In addition, I will change the data type from int to string because ids should not be numeric and they aren't intended to perform calculations.

**Code**

In [15]:
# rename to `tweet_id`
df_json = df_json.rename(columns={'id': 'tweet_id'})

In [16]:
# change `tweet_id` datatype to string in all 3 data frames
df_archive['tweet_id']= df_archive['tweet_id'].astype(str)
df_pred['tweet_id']= df_pred['tweet_id'].astype(str)
df_json['tweet_id']= df_json['tweet_id'].astype(str)

**Test**

In [17]:
df_json.head(1)

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467

In [18]:
df_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null object
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(2), object(1)
memory usage: 55.2+ KB


**Define**
2. Change **timestamp** data type in `df_archive`. This step will help if I was to analyze the data based on time or date.

**Code**

In [19]:
# change data type to datetime
df_archive['timestamp'] = pd.to_datetime(df_archive.timestamp)

**Test**

In [20]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](

**Define**
3. Drop **retweets** rows in `df_archive`. Only original tweets must be included in the final master dataset.

**Code**

In [21]:
# check how many retweets are in the dataframe
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](

There are 181 retweets in the dataframe. Next, I will delete the retweets.

In [22]:
# select retweets by checking the retweets columns (if they are empty it's an original tweet).
df_archive_notnull = df_archive[df_archive.retweeted_status_id.notnull()]
df_archive_notnull.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19  888202515573088257                    NaN                  NaN   
32  886054160059072513                    NaN                  NaN   
36  885311592912609280                    NaN                  NaN   
68  879130579576475649                    NaN                  NaN   
73  878404777348136964                    NaN                  NaN   

             timestamp                                             source  \
19 2017-07-21 01:02:36  <a href="http://twitter.com/download/iphone" r...   
32 2017-07-15 02:45:48  <a href="http://twitter.com/download/iphone" r...   
36 2017-07-13 01:35:06  <a href="http://twitter.com/download/iphone" r...   
68 2017-06-26 00:13:58  <a href="http://twitter.com/download/iphone" r...   
73 2017-06-24 00:09:53  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
19  RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32  RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36  RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68  RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73  RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   

    retweeted_status_user_id retweeted_status_timestamp  \
19              4.196984e+09  2017-07-19 00:47:34 +0000   
32              1.960740e+07  2017-07-15 02:44:07 +0000   
36              4.196984e+09  2017-02-12 01:04:29 +0000   
68              4.196984e+09  2017-06-23 01:10:23 +0000   
73              4.196984e+09  2017-06-23 16:00:04 +0000   

                                        expanded_urls  rating_numerator  \
19  https://twitter.com/dog_rates/status/887473957...                13   
32  https://twitter.com/dog_rates/status/886053434...                12   
36  https://twitter.com/dog_rates/status/830583320...                13   
68  https://twitter.com/dog_rates/status/878057613...                14   
73  https://www.gofundme.com/3yd6y1c,https://twitt...                13   

    rating_denominator    name doggo floofer pupper puppo  
19                  10  Canela  None    None   None  None  
32                  10    None  None    None   None  None  
36                  10   Lilly  None    None   None  None  
68                  10    Emmy  None    None   None  None  
73                  10  Shadow  None    None   None  None

In [23]:
# drop retweets (df_archive.retweeted_status rows that are not empty)
df_archive = df_archive.drop(df_archive_notnull.index, axis=0)
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

**Test**

In [24]:
# confirm changes (retweeted_status_id should be 0)
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null datetime64[ns]
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: datetime64[ns](1), fl

**Define**
4. Drop unnecessary columns in `df_archive`. It is not properly an issue, but it will make it easier to analyze the data. That is why I have 9 bullet points in this section.

**Code**

In [25]:
# drop columns
df_archive.drop(['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'source'], axis=1, inplace = True)

**Test**

In [26]:
# confirm changes
df_archive.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

**Define**
5. Drop null values from **expanded_urls** in `df_archive`. The **expanded_urls** is the only column that has a few missing values. I could not determine why that was the case nor I could fill these values out; therefore, I will drop these values.

**Code**

In [27]:
# drop null values
df_archive.dropna(inplace=True)

**Test**

In [28]:
# confirm changes
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2117 non-null object
timestamp             2117 non-null datetime64[ns]
text                  2117 non-null object
expanded_urls         2117 non-null object
rating_numerator      2117 non-null int64
rating_denominator    2117 non-null int64
name                  2117 non-null object
doggo                 2117 non-null object
floofer               2117 non-null object
pupper                2117 non-null object
puppo                 2117 non-null object
dtypes: datetime64[ns](1), int64(2), object(8)
memory usage: 198.5+ KB


**Define**
6. Drop tweets that are not dogs. This analysis should only incorporate tweets of dogs; therefore, the next step is to delete rows that contain non-dog tweets. In order to do that, I will merge the df_pred dataframe with df_archive and create a new dataframe df_archive2. The df_pred contains 3 predictions from the neural network whether a picture shows a dog. I assume that if any of the predictions suggest that it is a dog, I treated it as a dog. I checked a few tweets ‘by hand’ and it appeared to be true.

**Code**

In [29]:
# merge predictions into the archive
df_archive2 = df_archive.merge(df_pred, on=['tweet_id'], how='left')

In [30]:
# check if the dataframes were merged correctly
df_archive2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 22 columns):
tweet_id              2117 non-null object
timestamp             2117 non-null datetime64[ns]
text                  2117 non-null object
expanded_urls         2117 non-null object
rating_numerator      2117 non-null int64
rating_denominator    2117 non-null int64
name                  2117 non-null object
doggo                 2117 non-null object
floofer               2117 non-null object
pupper                2117 non-null object
puppo                 2117 non-null object
jpg_url               1994 non-null object
img_num               1994 non-null float64
p1                    1994 non-null object
p1_conf               1994 non-null float64
p1_dog                1994 non-null object
p2                    1994 non-null object
p2_conf               1994 non-null float64
p2_dog                1994 non-null object
p3                    1994 non-null object
p3_conf               19

In [31]:
# select pictures if any of the predictions suggest that it is a dog
df_archive2 = df_archive2.query('p1_dog == True | p2_dog == True | p3_dog == True')

**Test**

In [32]:
# check how many dog tweets are left
df_archive2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1686 entries, 1 to 2116
Data columns (total 22 columns):
tweet_id              1686 non-null object
timestamp             1686 non-null datetime64[ns]
text                  1686 non-null object
expanded_urls         1686 non-null object
rating_numerator      1686 non-null int64
rating_denominator    1686 non-null int64
name                  1686 non-null object
doggo                 1686 non-null object
floofer               1686 non-null object
pupper                1686 non-null object
puppo                 1686 non-null object
jpg_url               1686 non-null object
img_num               1686 non-null float64
p1                    1686 non-null object
p1_conf               1686 non-null float64
p1_dog                1686 non-null object
p2                    1686 non-null object
p2_conf               1686 non-null float64
p2_dog                1686 non-null object
p3                    1686 non-null object
p3_conf               16

**Define**
7. Clean **rating_denominator** in `df_archive`. Extract the values from text, delete tweets without ratings and tweets with multiple dogs. Change the data type to float.

8. Clean **rating_denominator** in `df_archive`. Extract the values from text, delete tweets without ratings and tweets with multiple dogs. Change the data type to float.

**Code**

In [33]:
# change column width to view full text
pd.options.display.max_colwidth = 200

In [34]:
# import regular expression
import re

In [35]:
# extract ratings from the text column into a new column 'ratings'.
# Some ratings need to be fixed manually because there are multiple xx/xx occurrences in text
# or there are multiple dogs in one picture
df_archive2['ratings'] = df_archive2['text'].str.extract(r'(\d+\.?\d*/\d+){1}', expand = False)

In [36]:
# Extract rating numerators from the new column 'ratings' as float
df_archive2['rating_numerator'] = df_archive2['ratings'].str.split('/').str.get(0).astype(float)

In [37]:
# Extract rating denominators from the new column 'ratings' as float
df_archive2['rating_denominator'] = df_archive2['ratings'].str.split('/').str.get(1).astype(float)

In [38]:
# drop ratings column after the extraction
df_archive2.drop(['ratings'], axis=1, inplace=True)

In this section, I'm going to manually clean incorrecntly extracted ratings. I will delete tweets with multiple dogs.

In [39]:
df_archive2['rating_denominator'].value_counts()

10.0     1669
50.0        3
80.0        2
11.0        2
130.0       1
150.0       1
2.0         1
120.0       1
110.0       1
40.0        1
90.0        1
20.0        1
7.0         1
70.0        1
Name: rating_denominator, dtype: int64

In [40]:
df_archive2[df_archive2.rating_denominator == 70]

tweet_id           timestamp  \
341  820690176645140481 2017-01-15 17:52:40   

                                                                                                    text  \
341  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   

                                                                                                                                                                                       expanded_urls  \
341  https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1   

     rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
341              84.0                70.0  None  None    None   None  ...     

    img_num                           p1   p1_conf p1_dog      p2   p2_conf  \
341     2.0  West_Highland_white_terrier  0.872064   True  kuvasz  0.059526   

    p2_dog       p3 p3_conf p3_dog  
341   True  Samoyed  0.0374   True  

[1 rows x 22 columns]

In [41]:
# drop row 341 because it is multiple dogs
df_archive2.drop(341, inplace=True)

In [42]:
df_archive2[df_archive2.rating_denominator == 7]

tweet_id           timestamp  \
405  810984652412424192 2016-12-19 23:06:23   

                                                                                                                                                                          text  \
405  Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   

                                                                                           expanded_urls  \
405  https://www.gofundme.com/sams-smile,https://twitter.com/dog_rates/status/810984652412424192/photo/1   

     rating_numerator  rating_denominator name doggo floofer pupper  ...    \
405              24.0                 7.0  Sam  None    None   None  ...     

    img_num                p1   p1_conf p1_dog               p2   p2_conf  \
405     1.0  golden_retriever  0.871342   True  Tibetan_mastiff  0.036708   

    p2_dog                  p3   p3_conf p3_dog  
405   True  Labrador_retriever  0.025823   True  

[1 rows x 22 columns]

In [43]:
# drop row 405 as there is no rating provided
df_archive2.drop(405, inplace=True)

In [44]:
df_archive2[df_archive2.rating_denominator == 20]

tweet_id           timestamp  \
953  722974582966214656 2016-04-21 02:25:47   

                                                                 text  \
953  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   

                                                       expanded_urls  \
953  https://twitter.com/dog_rates/status/722974582966214656/photo/1   

     rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
953               4.0                20.0  None  None    None   None  ...     

    img_num          p1   p1_conf p1_dog                          p2  \
953     1.0  Great_Dane  0.246762   True  Greater_Swiss_Mountain_dog   

      p2_conf p2_dog          p3   p3_conf p3_dog  
953  0.126131   True  Weimaraner  0.085297   True  

[1 rows x 22 columns]

In [45]:
df_archive2.loc[953, 'rating_denominator'] = 10
df_archive2.loc[953, 'rating_numerator'] = 13

In [46]:
df_archive2[df_archive2.rating_denominator == 40]

tweet_id           timestamp  \
1218  697463031882764288 2016-02-10 16:51:59   

                                                                                              text  \
1218  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   

                                                        expanded_urls  \
1218  https://twitter.com/dog_rates/status/697463031882764288/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1218              44.0                40.0  None  None    None   None  ...     

     img_num                  p1   p1_conf p1_dog                p2   p2_conf  \
1218     1.0  Labrador_retriever  0.999885   True  golden_retriever  0.000098   

     p2_dog   p3   p3_conf p3_dog  
1218   True  pug  0.000008   True  

[1 rows x 22 columns]

In [47]:
# drop row 1218 because it is multiple dogs
df_archive2.drop(1218, inplace=True)

In [48]:
df_archive2[df_archive2.rating_denominator == 90]

tweet_id           timestamp  \
1016  713900603437621249 2016-03-27 01:29:02   

                                                                                               text  \
1016  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   

                                                        expanded_urls  \
1016  https://twitter.com/dog_rates/status/713900603437621249/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1016              99.0                90.0  None  None    None   None  ...     

     img_num                p1   p1_conf p1_dog              p2   p2_conf  \
1016     1.0  golden_retriever  0.371816   True  cocker_spaniel  0.177413   

     p2_dog            p3   p3_conf p3_dog  
1016   True  Irish_setter  0.092725   True  

[1 rows x 22 columns]

In [49]:
# drop row 1016 because it is multiple dogs
df_archive2.drop(1016, inplace=True)

In [50]:
df_archive2[df_archive2.rating_denominator == 110]

tweet_id           timestamp  \
1411  684222868335505415 2016-01-05 04:00:18   

                                                                                                                                              text  \
1411  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   

                                                        expanded_urls  \
1411  https://twitter.com/dog_rates/status/684222868335505415/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1411             121.0               110.0  None  None    None   None  ...     

     img_num                           p1   p1_conf p1_dog              p2  \
1411     1.0  soft-coated_wheaten_terrier  0.791182   True  cocker_spaniel   

       p2_conf p2_dog     p3   p3_conf p3_dog  
1411  0.072444   True  teddy  0.071486  False  

[1 rows x 22 columns]

In [51]:
# drop row 1411 because it is multiple dogs
df_archive2.drop(1411, inplace=True)

In [52]:
df_archive2[df_archive2.rating_denominator == 120]

tweet_id           timestamp  \
1552  677716515794329600 2015-12-18 05:06:23   

                                                                        text  \
1552  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq   

                                                        expanded_urls  \
1552  https://twitter.com/dog_rates/status/677716515794329600/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1552             144.0               120.0  None  None    None   None  ...     

     img_num     p1   p1_conf p1_dog    p2   p2_conf p2_dog    p3   p3_conf  \
1552     1.0  teddy  0.662908  False  crib  0.031891  False  chow  0.025438   

     p3_dog  
1552   True  

[1 rows x 22 columns]

In [53]:
# drop row 1552 because it is multiple dogs
df_archive2.drop(1552, inplace=True)

In [54]:
df_archive2[df_archive2.rating_denominator == 130]

tweet_id           timestamp  \
1410  684225744407494656 2016-01-05 04:11:44   

                                                                                                                                text  \
1410  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3   

                                                                                                                        expanded_urls  \
1410  https://twitter.com/dog_rates/status/684225744407494656/photo/1,https://twitter.com/dog_rates/status/684225744407494656/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1410             143.0               130.0  None  None    None   None  ...     

     img_num                p1   p1_conf p1_dog       p2   p2_conf p2_dog  \
1410     2.0  golden_retriever  0.203249   True  Samoyed  0.067958   True   

                  p3   p3_conf p3_dog  
1410  Great_Pyrenees  0.065327   True  

[1 rows x 22 columns]

In [55]:
# drop row 1410 because it is multiple dogs
df_archive2.drop(1410, inplace=True)

In [56]:
df_archive2[df_archive2.rating_denominator == 150]

tweet_id           timestamp  \
702  758467244762497024 2016-07-28 01:00:57   

                                                                               text  \
702  Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   

                                                       expanded_urls  \
702  https://twitter.com/dog_rates/status/758467244762497024/video/1   

     rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
702             165.0               150.0  None  None    None   None  ...     

    img_num                  p1   p1_conf p1_dog         p2   p2_conf p2_dog  \
702     1.0  Labrador_retriever  0.436377   True  Chihuahua  0.113956   True   

                                 p3   p3_conf p3_dog  
702  American_Staffordshire_terrier  0.099689   True  

[1 rows x 22 columns]

In [57]:
# drop row 1410 because it is multiple dogs
df_archive2.drop(702, inplace=True)

In [58]:
df_archive2[df_archive2.rating_denominator == 2]

tweet_id           timestamp  \
2096  666287406224695296 2015-11-16 16:11:11   

                                                                                                                                          text  \
2096  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

                                                        expanded_urls  \
2096  https://twitter.com/dog_rates/status/666287406224695296/photo/1   

      rating_numerator  rating_denominator name doggo floofer pupper  ...    \
2096               1.0                 2.0   an  None    None   None  ...     

     img_num           p1   p1_conf p1_dog          p2   p2_conf p2_dog  \
2096     1.0  Maltese_dog  0.857531   True  toy_poodle  0.063064   True   

                    p3   p3_conf p3_dog  
2096  miniature_poodle  0.025581   True  

[1 rows x 22 columns]

In [59]:
# change row 2096
df_archive2.loc[2096, 'rating_denominator'] = 10
df_archive2.loc[2096, 'rating_numerator'] = 9

In [60]:
df_archive2[df_archive2.rating_denominator == 11]

tweet_id           timestamp  \
857   740373189193256964 2016-06-08 02:41:38   
1438  682962037429899265 2016-01-01 16:30:13   

                                                                                                                                              text  \
857   After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1438   This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   

                                                                                                                                                                                                expanded_urls  \
857   https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,http...   
1438                                                                                                                                          https://twitter.com/dog_rates/status/682962037429899265/photo/1   

      rating_numerator  rating_denominator    name doggo floofer pupper  \
857                9.0                11.0    None  None    None   None   
1438               7.0                11.0  Darrel  None    None   None   

      ...   img_num                p1   p1_conf p1_dog         p2   p2_conf  \
857   ...       3.0  golden_retriever  0.807644   True     kuvasz  0.101286   
1438  ...       1.0             dingo  0.278600  False  Chihuahua  0.155207   

     p2_dog                  p3   p3_conf p3_dog  
857    True  Labrador_retriever  0.023785   True  
1438   True               loupe  0.153598  False  

[2 rows x 22 columns]

In [61]:
df_archive2.loc[857, 'rating_denominator'] = 10
df_archive2.loc[857, 'rating_numerator'] = 14

In [62]:
df_archive2.loc[1438, 'rating_denominator'] = 10
df_archive2.loc[1438, 'rating_numerator'] = 10

In [63]:
df_archive2[df_archive2.rating_denominator == 80]

tweet_id           timestamp  \
1041  710658690886586372 2016-03-18 02:46:49   
1615  675853064436391936 2015-12-13 01:41:41   

                                                                                                              text  \
1041  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1615   Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw   

                                                                                                                        expanded_urls  \
1041                                                                  https://twitter.com/dog_rates/status/710658690886586372/photo/1   
1615  https://twitter.com/dog_rates/status/675853064436391936/photo/1,https://twitter.com/dog_rates/status/675853064436391936/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
1041              80.0                80.0  None  None    None   None  ...     
1615              88.0                80.0  None  None    None   None  ...     

     img_num                           p1   p1_conf p1_dog                p2  \
1041     1.0  soft-coated_wheaten_terrier  0.948617   True    Dandie_Dinmont   
1615     1.0           Labrador_retriever  0.868367   True  golden_retriever   

       p2_conf p2_dog      p3   p3_conf p3_dog  
1041  0.018664   True   cairn  0.015943   True  
1615  0.043305   True  vizsla  0.028207   True  

[2 rows x 22 columns]

In [64]:
# drop rows 1041 and 1615 because it is multiple dogs
df_archive2.drop([1041,1615], inplace=True)

In [65]:
df_archive2[df_archive2.rating_denominator == 50]

tweet_id           timestamp  \
990   716439118184652801 2016-04-03 01:36:11   
1061  709198395643068416 2016-03-14 02:04:08   
1136  704054845121142784 2016-02-28 21:25:30   

                                                                                                                                                    text  \
990                                This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1061  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1136                                                                  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   

                                                        expanded_urls  \
990   https://twitter.com/dog_rates/status/716439118184652801/photo/1   
1061  https://twitter.com/dog_rates/status/709198395643068416/photo/1   
1136  https://twitter.com/dog_rates/status/704054845121142784/photo/1   

      rating_numerator  rating_denominator      name doggo floofer pupper  \
990               50.0                50.0  Bluebert  None    None   None   
1061              45.0                50.0      None  None    None   None   
1136              60.0                50.0         a  None    None   None   

      ...   img_num              p1   p1_conf p1_dog                       p2  \
990   ...       1.0  Siberian_husky  0.396495   True                 malamute   
1061  ...       1.0          borzoi  0.490783   True  wire-haired_fox_terrier   
1136  ...       1.0  Great_Pyrenees  0.667939   True                   kuvasz   

       p2_conf p2_dog                p3   p3_conf p3_dog  
990   0.317053   True        Eskimo_dog  0.273419   True  
1061  0.083513   True    English_setter  0.083184   True  
1136  0.228764   True  golden_retriever  0.043885   True  

[3 rows x 22 columns]

In [66]:
df_archive2.loc[990, 'rating_denominator'] = 10
df_archive2.loc[990, 'rating_numerator'] = 11

In [67]:
# drop rows 1061 and 1136 because it is multiple dogs
df_archive2.drop([1061,1136], inplace=True)

**Clean Rating Numerator**

In [68]:
df_archive2.rating_numerator.value_counts()

12.00    423
10.00    361
11.00    357
13.00    230
9.00     133
8.00      68
7.00      31
14.00     24
6.00      16
5.00      13
4.00       6
3.00       5
2.00       2
13.50      1
11.27      1
0.00       1
11.26      1
9.75       1
Name: rating_numerator, dtype: int64

In [69]:
df_archive2[df_archive2.rating_numerator == 0]

tweet_id           timestamp  \
249  835152434251116546 2017-02-24 15:40:31   

                                                                                                                  text  \
249  When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag   

                                                                                                                                                                                       expanded_urls  \
249  https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1   

     rating_numerator  rating_denominator  name doggo floofer pupper  ...    \
249               0.0                10.0  None  None    None   None  ...     

    img_num     p1   p1_conf p1_dog                              p2   p2_conf  \
249     3.0  swing  0.967066  False  American_Staffordshire_terrier  0.012731   

    p2_dog                         p3   p3_conf p3_dog  
249   True  Staffordshire_bullterrier  0.007039   True  

[1 rows x 22 columns]

In [70]:
df_archive2[df_archive2.rating_numerator == 2]

tweet_id           timestamp  \
1538  678424312106393600 2015-12-20 03:58:55   
2010  667878741721415682 2015-11-21 01:34:35   

                                                                                                                                              text  \
1538  This is Crystal. She's a shitty fireman. No sense of urgency. People could be dying Crystal. 2/10 just irresponsible https://t.co/rtMtjSl9pz   
2010   This is Tedrick. He lives on the edge. Needs someone to hit the gas tho. Other than that he's a baller. 10&amp;2/10 https://t.co/LvP1TTYSCN   

                                                        expanded_urls  \
1538  https://twitter.com/dog_rates/status/678424312106393600/photo/1   
2010  https://twitter.com/dog_rates/status/667878741721415682/photo/1   

      rating_numerator  rating_denominator     name doggo floofer pupper  \
1538               2.0                10.0  Crystal  None    None   None   
2010               2.0                10.0  Tedrick  None    None   None   

      ...   img_num           p1   p1_conf p1_dog                  p2  \
1538  ...       1.0  Maltese_dog  0.759945   True          toy_poodle   
2010  ...       1.0    seat_belt  0.200373  False  miniature_pinscher   

       p2_conf p2_dog          p3   p3_conf p3_dog  
1538  0.101194   True    Shih-Tzu  0.056037   True  
2010  0.106003   True  schipperke  0.104733   True  

[2 rows x 22 columns]

In [71]:
df_archive2[df_archive2.rating_numerator == 3]

tweet_id           timestamp  \
589   777885040357281792 2016-09-19 15:00:20   
977   718246886998687744 2016-04-08 01:19:36   
1036  711306686208872448 2016-03-19 21:41:44   
1706  673906403526995968 2015-12-07 16:46:21   
1948  668989615043424256 2015-11-24 03:08:48   

                                                                                                                                              text  \
589   This is Wesley. He's clearly trespassing. Seems rather h*ckin violent too. Weaponized forehead. 3/10 wouldn't let in https://t.co/pL7wbMRW7M   
977           This is Alexanderson. He's got a weird ass birth mark. Dreadful at fetch. Won't eat kibble. 3/10 wtf @Target https://t.co/FmxOpf2Sgl   
1036     What hooligan sent in pictures w/out a dog in them? Churlish af. 3/10 just bc that's a neat fluffy bean bag chair https://t.co/wcwoGOkZvz   
1706   Guys I'm getting real tired of this. We only rate dogs. Please don't send in other things like this Bulbasaur. 3/10 https://t.co/t5rQHl6W8M   
1948  This is Bernie. He's taking his Halloween costume very seriously. Wants to be baked. 3/10 not a good idea Bernie smh https://t.co/1zBp1moFlX   

                                                                                                                        expanded_urls  \
589   https://twitter.com/dog_rates/status/777885040357281792/photo/1,https://twitter.com/dog_rates/status/777885040357281792/photo/1   
977                                                                   https://twitter.com/dog_rates/status/718246886998687744/photo/1   
1036  https://twitter.com/dog_rates/status/711306686208872448/photo/1,https://twitter.com/dog_rates/status/711306686208872448/photo/1   
1706                                                                  https://twitter.com/dog_rates/status/673906403526995968/photo/1   
1948                                                                  https://twitter.com/dog_rates/status/668989615043424256/photo/1   

      rating_numerator  rating_denominator          name doggo floofer pupper  \
589                3.0                10.0        Wesley  None    None   None   
977                3.0                10.0  Alexanderson  None    None   None   
1036               3.0                10.0          None  None    None   None   
1706               3.0                10.0          None  None    None   None   
1948               3.0                10.0        Bernie  None    None   None   

      ...   img_num                  p1   p1_conf p1_dog  \
589   ...       1.0        Afghan_hound  0.123529   True   
977   ...       1.0           Chihuahua  0.354488   True   
1036  ...       1.0  leatherback_turtle  0.280835  False   
1706  ...       1.0         toilet_seat  0.683319  False   
1948  ...       1.0                 pug  0.917326   True   

                               p2   p2_conf p2_dog              p3   p3_conf  \
589                        basset  0.119682   True  Siberian_husky  0.108709   
977                        carton  0.159672  False  Siberian_husky  0.057498   
1036                   loggerhead  0.123290  False  Dandie_Dinmont  0.086792   
1706  soft-coated_wheaten_terrier  0.048928   True  Siberian_husky  0.030386   
1948                  waffle_iron  0.014918  False       Chihuahua  0.013524   

     p3_dog  
589    True  
977    True  
1036   True  
1706   True  
1948   True  

[5 rows x 22 columns]

In [72]:
df_archive2[df_archive2.rating_numerator == 4]

tweet_id           timestamp  \
797   747816857231626240 2016-06-28 15:40:07   
1089  707420581654872064 2016-03-09 04:19:44   
1475  680940246314430465 2015-12-27 02:36:20   
1836  671122204919246848 2015-11-30 00:22:57   
2050  667176164155375616 2015-11-19 03:02:47   
2077  666649482315059201 2015-11-17 16:09:56   

                                                                                                                                              text  \
797     Viewer discretion is advised. This is a terrible attack in progress. Not even in water (tragic af). 4/10 bad sherk https://t.co/L3U0j14N5R   
1089  This is Keurig. He's a rare dog. Laughs like an idiot tho. Head is basically a weapon. Poorly maintained goatee 4/10 https://t.co/xOrUyj7K30   
1475                                                                                   This is Alice. She's an idiot. 4/10 https://t.co/VQXdwJfkyS   
1836      Two miniature golden retrievers here. Webbed paws. Don't walk very efficiently. Can't catch a tennis ball. 4/10s https://t.co/WzVLdSHJU7   
2050  These are strange dogs. All have toupees. Long neck for dogs. In a shed of sorts? Work in groups? 4/10 still petable https://t.co/PZxSarAfSN   
2077    Cool dog. Enjoys couch. Low monotone bark. Very nice kicks. Pisses milk (must be rare). Can't go down stairs. 4/10 https://t.co/vXMKrJC81s   

                                                        expanded_urls  \
797   https://twitter.com/dog_rates/status/747816857231626240/photo/1   
1089  https://twitter.com/dog_rates/status/707420581654872064/photo/1   
1475  https://twitter.com/dog_rates/status/680940246314430465/photo/1   
1836  https://twitter.com/dog_rates/status/671122204919246848/photo/1   
2050  https://twitter.com/dog_rates/status/667176164155375616/photo/1   
2077  https://twitter.com/dog_rates/status/666649482315059201/photo/1   

      rating_numerator  rating_denominator    name doggo floofer pupper  \
797                4.0                10.0       a  None    None   None   
1089               4.0                10.0  Keurig  None    None   None   
1475               4.0                10.0   Alice  None    None   None   
1836               4.0                10.0    None  None    None   None   
2050               4.0                10.0    None  None    None   None   
2077               4.0                10.0    None  None    None   None   

      ...   img_num                           p1   p1_conf p1_dog  \
797   ...       1.0                     Pembroke  0.768923   True   
1089  ...       1.0                          ram  0.518215  False   
1475  ...       1.0  soft-coated_wheaten_terrier  0.289598   True   
1836  ...       1.0                        goose  0.351957  False   
2050  ...       1.0  soft-coated_wheaten_terrier  0.318981   True   
2077  ...       1.0                Border_collie  0.447803   True   

                               p2   p2_conf p2_dog                 p3  \
797                     Chihuahua  0.029053   True  Shetland_sheepdog   
1089                       kuvasz  0.149391   True     Great_Pyrenees   
1475  West_Highland_white_terrier  0.157195   True         toy_poodle   
1836                    Chihuahua  0.101228   True                hen   
2050             Lakeland_terrier  0.215218   True         toy_poodle   
2077             English_springer  0.170497   True             collie   

       p3_conf p3_dog  
797   0.029035   True  
1089  0.106003   True  
1475  0.074435   True  
1836  0.065818  False  
2050  0.106014   True  
2077  0.139206   True  

[6 rows x 22 columns]

In [73]:
df_archive2[df_archive2.rating_numerator == 5]

tweet_id           timestamp  \
561   781661882474196992 2016-09-30 01:08:10   
750   751583847268179968 2016-07-09 01:08:47   
1288  691483041324204033 2016-01-25 04:49:38   
1421  683849932751646720 2016-01-04 03:18:23   
1569  677301033169788928 2015-12-17 01:35:24   
1592  676588346097852417 2015-12-15 02:23:26   
1671  674646392044941312 2015-12-09 17:46:48   
1746  672980819271634944 2015-12-05 03:28:25   
1858  670782429121134593 2015-11-29 01:52:48   
1946  668994913074286592 2015-11-24 03:29:51   
1970  668631377374486528 2015-11-23 03:25:17   
2073  666776908487630848 2015-11-18 00:36:17   
2112  666049248165822465 2015-11-16 00:24:50   

                                                                                                                                                  text  \
561                         Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC   
750           Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8   
1288  When bae says they can't go out but you see them with someone else that same night. 5/10 &amp; 10/10 for heartbroken pup https://t.co/aenk0KpoWM   
1421                               This is Jiminy. He's not the brightest dog. Needs to lay off the kibble. 5/10 still petable https://t.co/omln4LOy1x   
1569                                               This is Juckson. He's totally on his way to a nascar race. 5/10 for Juckson https://t.co/IoLRvF0Kak   
1592      This is Bubbles. He kinda resembles a fish. Always makes eye contact with u no matter what. Sneaky tongue slip. 5/10 https://t.co/Nrhvc5tLFT   
1671     Two gorgeous dogs here. Little waddling dog is a rebel. Refuses to look at camera. Must be a preteen. 5/10 &amp; 8/10 https://t.co/YPfw7oahbD   
1746                         Extraordinary dog here. Looks large. Just a head. No body. Rather intrusive. 5/10 would still pet https://t.co/ufHWUFA9Pu   
1858                                                                     This dude slaps your girl's ass what do you do?\n5/10 https://t.co/6dioUL6gcP   
1946      Two gorgeous pups here. Both have cute fake horns(adorable). Barn in the back looks on fire. 5/10 would pet rly well https://t.co/w5oYFXi0uh   
1970           Meet Zeek. He is a grey Cumulonimbus. Zeek is hungry. Someone should feed Zeek asap. 5/10 absolutely terrifying https://t.co/fvVNScw8VH   
2073       This is Josep. He is a Rye Manganese mix. Can drive w eyes closed. Very irresponsible. Menace on the roadways. 5/10 https://t.co/XNGeDwrtYH   
2112                          Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq   

                                                                                                                                                                                                expanded_urls  \
561                                                                                                                                           https://twitter.com/dog_rates/status/781661882474196992/photo/1   
750                                                                                                                                           https://twitter.com/dog_rates/status/751583847268179968/photo/1   
1288  https://twitter.com/dog_rates/status/691483041324204033/photo/1,https://twitter.com/dog_rates/status/691483041324204033/photo/1,https://twitter.com/dog_rates/status/691483041324204033/photo/1,http...   
1421                                                                                                                                          https://twitter.com/dog_rates/status/683849932751646720/photo/1   
1569                                                                                                                                         

In [74]:
df_archive2[df_archive2.rating_numerator == 6]

tweet_id           timestamp  \
186   848212111729840128 2017-04-01 16:35:01   
786   748575535303884801 2016-06-30 17:54:50   
867   739485634323156992 2016-06-05 15:54:48   
1173  700462010979500032 2016-02-18 23:28:52   
1243  695051054296211456 2016-02-04 01:07:39   
1533  678740035362037760 2015-12-21 00:53:29   
1560  677565715327688705 2015-12-17 19:07:09   
1611  675891555769696257 2015-12-13 04:14:39   
1739  673240798075449344 2015-12-05 20:41:29   
1837  671115716440031232 2015-11-29 23:57:10   
1866  670704688707301377 2015-11-28 20:43:53   
1928  669375718304980992 2015-11-25 04:43:02   
2002  668113020489474048 2015-11-21 17:05:31   
2083  666430724426358785 2015-11-17 01:40:41   
2103  666082916733198337 2015-11-16 02:38:37   
2113  666044226329800704 2015-11-16 00:04:52   

                                                                                                                                                                    text  \
186   This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1   
786                         This is one of the most reckless puppers I've ever seen. How she got a license in the first place is beyond me. 6/10 https://t.co/z5bAdtn9kd   
867                            This is Kyle. He's a heavy drinker and an avid pot user. Just wants to be pupular. 6/10 I can't support this Kyle https://t.co/rRULp7XFnO   
1173                          This is Murphy. He's a mini golden retriever. Missing two legs (tragic). Mouth sharp. Looks rather perturbed. 6/10 https://t.co/ALO02IAKCn   
1243                        Meet Brian (pronounced "Kirk"). He's not amused by ur churlish tomfoolery. Once u put him down you're done for. 6/10 https://t.co/vityMwPKKi   
1533                           Meet Tango. He's a large dog. Doesn't care much for personal space. Owner isn't very accepting. Tongue slip. 6/10 https://t.co/p2T5kGebxe   
1560                        Contortionist pup here. Inside pentagram. Clearly worships Satan. Known to slowly push fragile stuff off tables 6/10 https://t.co/EX9oR55VMe   
1611                         This is Donny. He's summoning the demon monster Babadook. 6/10 Donny please no that won't be a good time for anyone https://t.co/kiW6Knb7Gp   
1739                        Magical floating dog here. Very calm. Always hangs by the pond. Rather moist. Good listener. 6/10 personally I'd pet https://t.co/1euKoOvy49   
1837                        Meet Phred. He isn't steering, looking at the road, or wearing a seatbelt. Phred is a rolling tornado of danger 6/10 https://t.co/mZD7Bo7HfV   
1866                             Meet Danny. He's too good to look at the road when he's driving. Absolute menace. 6/10 completely irresponsible https://t.co/I1lMUy1FqH   
1928                                                     This is Billl. He's trying to be a ghost but he's not very good at it. 6/10 c'mon Billl https://t.co/ero0XfdGtY   
2002                                This is Alfie. He's that one hypocritical gym teacher who made you run laps. Great posture. Cool bench. 6/10 https://t.co/GCJzm3YsfX   
2083                                   Oh boy what a pup! Sunglasses take this one to the next level. Weirdly folds front legs. Pretty big. 6/10 https://t.co/yECbFrSArM   
2103                                       Here we have a well-established sunblockerspaniel. Lost his other flip-flop. 6/10 not very waterproof https://t.co/3RU6x0vHB7   
2113                           This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx   

                                                                                                                        expanded_urls  \
186                                                                   https://twitter.com/dog_rates/status/848212111729840128/photo/1   


In [75]:
df_archive2[df_archive2.rating_numerator == 7]

tweet_id           timestamp  \
696   759099523532779520 2016-07-29 18:53:24   
819   745712589599014916 2016-06-22 20:18:30   
1111  706166467411222528 2016-03-05 17:16:20   
1200  698703483621523456 2016-02-14 03:01:06   
1231  696405997980676096 2016-02-07 18:51:43   
1352  687494652870668288 2016-01-14 04:41:12   
1358  687312378585812992 2016-01-13 16:36:55   
1385  685641971164143616 2016-01-09 01:59:19   
1504  679854723806179328 2015-12-24 02:42:51   
1511  679530280114372609 2015-12-23 05:13:38   
1513  679511351870550016 2015-12-23 03:58:25   
1528  678798276842360832 2015-12-21 04:44:55   
1643  675146535592706048 2015-12-11 02:54:12   
1723  673612854080196609 2015-12-06 21:19:54   
1732  673345638550134785 2015-12-06 03:38:05   
1765  672538107540070400 2015-12-03 22:09:14   
1779  672239279297454080 2015-12-03 02:21:48   
1789  672068090318987265 2015-12-02 15:01:33   
1843  670833812859932673 2015-11-29 05:16:59   
1854  670792680469889025 2015-11-29 02:33:32   
1859  670780561024270336 2015-11-29 01:45:22   
1962  668815180734689280 2015-11-23 15:35:39   
1999  668171859951755264 2015-11-21 20:59:20   
2034  667491009379606528 2015-11-19 23:53:52   
2036  667455448082227200 2015-11-19 21:32:34   
2041  667405339315146752 2015-11-19 18:13:27   
2058  667090893657276420 2015-11-18 21:23:57   
2081  666437273139982337 2015-11-17 02:06:42   
2084  666428276349472768 2015-11-17 01:30:57   
2088  666407126856765440 2015-11-17 00:06:54   
2115  666029285002620928 2015-11-15 23:05:30   

                                                                                                                                              text  \
696   Meet Toby. He has a drinking problem. Inflatable marijuana plant in the back is also not a good look. 7/10 cmon Toby https://t.co/Cim4DSj6Oi   
819                      This is Percy. He fell asleep at the wheel. Irresponsible af. 7/10 absolute menace on the roadway https://t.co/QHbvtvaw8E   
1111         This is Dexter. He's a shy pup. Doesn't bark much. Dreadful fetcher. Has rare sun allergy. 7/10 still petable https://t.co/sA7P3JSqiv   
1200                          This is Rusty. He has no respect for POULTRY products. Unbelievable af. 7/10 would still pet https://t.co/hEH19t1eFp   
1231   This is Berb. He just found out that they have made 31 Kidz Bop CD's. Downright terrifying. 7/10 hang in there Berb https://t.co/CIFLjiTFwZ   
1352                                                                       This is Marq. He stole this car. 7/10 wtf Marq? https://t.co/MHScqo5l8c   
1358                                            "You got any games on your phone" 7/10 for invasive brown Dalmatian pupper https://t.co/yzGR9xjE9Q   
1385                                  This is Otis. He just passed a cop while going 61 in a 45. Very nervous pupper. 7/10 https://t.co/jJS8qQeuNO   
1504  This is Bruce. He's a rare pup. Covered in Frosted Flakes. Nifty gold teeth. Overall good dog. 7/10 would pet firmly https://t.co/RtxxACzZ8A   
1511  Guys this really needs to stop. We've been over this way too many times. This is a giraffe. We only rate dogs.. 7/10 https://t.co/yavgkHYPOC   
1513         Say hello to William. He makes fun of others because he's terrified of his own deep-seated insecurities. 7/10 https://t.co/bwuV6FlRxr   
1528                                                                          This is Linda. She fucking hates trees. 7/10 https://t.co/blaY85FIxR   
1643                                            This is Coops. He's yelling at the carpet. Not very productive Coops. 7/10 https://t.co/Uz52oYnHzF   
1723       This is Jeffri. He's a speckled ice pupper. Very lazy. Enjoys the occasional swim. Rather majestic really. 7/10 https://t.co/0iyItbtkr8   
1732      This is Gerald. He's a fluffy lil yellow pup. Always looks like his favorite team just lost on a hail mary. 7/10 https://t.co/GpSkpN8kXS   
1765                                              Meet Darby. He's a Fiscal Tut

In [76]:
df_archive2[df_archive2.rating_numerator == 13.5]

tweet_id           timestamp  \
41  883482846933004288 2017-07-08 00:28:19   

                                                                                                                                        text  \
41  This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   

                                                                                                                      expanded_urls  \
41  https://twitter.com/dog_rates/status/883482846933004288/photo/1,https://twitter.com/dog_rates/status/883482846933004288/photo/1   

    rating_numerator  rating_denominator   name doggo floofer pupper  ...    \
41              13.5                10.0  Bella  None    None   None  ...     

   img_num                p1   p1_conf p1_dog                  p2   p2_conf  \
41     1.0  golden_retriever  0.943082   True  Labrador_retriever  0.032409   

   p2_dog      p3   p3_conf p3_dog  
41   True  kuvasz  0.005501   True  

[1 rows x 22 columns]

**Test**

In [77]:
# confirm that all denominators are 10
df_archive2['rating_denominator'].value_counts()

10.0    1674
Name: rating_denominator, dtype: int64

In [78]:
# check for changed data type (float for ratings)
df_archive2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1674 entries, 1 to 2116
Data columns (total 22 columns):
tweet_id              1674 non-null object
timestamp             1674 non-null datetime64[ns]
text                  1674 non-null object
expanded_urls         1674 non-null object
rating_numerator      1674 non-null float64
rating_denominator    1674 non-null float64
name                  1674 non-null object
doggo                 1674 non-null object
floofer               1674 non-null object
pupper                1674 non-null object
puppo                 1674 non-null object
jpg_url               1674 non-null object
img_num               1674 non-null float64
p1                    1674 non-null object
p1_conf               1674 non-null float64
p1_dog                1674 non-null object
p2                    1674 non-null object
p2_conf               1674 non-null float64
p2_dog                1674 non-null object
p3                    1674 non-null object
p3_conf             

In [79]:
# confirm changes
df_archive2.rating_numerator.value_counts()

12.00    423
10.00    361
11.00    357
13.00    230
9.00     133
8.00      68
7.00      31
14.00     24
6.00      16
5.00      13
4.00       6
3.00       5
2.00       2
13.50      1
11.27      1
0.00       1
11.26      1
9.75       1
Name: rating_numerator, dtype: int64

All rating numerators look clean.

**Define**
8. Clean names using regular expression in `df_archive`. A useful pattern to find the incorrect names is that they started with lowercase, instead, the true names always started with uppercase. In addition, the real names usually are placed after words like "named", "is" etc.

**Code**

In [80]:
# extract correct names from text
df_archive2['name'] = df_archive2['text'].str.extract(r'([is|named|to|meet]+\s[A-Z][\W]?[A-Z]?[a-z]+)', expand = False).str.split(' ').str.get(1)

In [81]:
# Change 'I' to NaN
df_archive2.loc[df_archive2['name'] == "I", 'name'] = np.nan

**Test**

In [82]:
df_archive2['name'].value_counts()

Charlie      10
Lucy         10
Cooper       10
Tucker        9
Oliver        9
Penny         8
Daisy         7
Winston       7
Toby          7
Sadie         7
Bo            6
Jax           6
Lola          6
Koda          6
Stanley       5
Reggie        5
Leo           5
Rusty         5
Zoey          5
Bella         5
Oscar         5
Boomer        4
Larry         4
Dave          4
Brody         4
Maggie        4
Scooter       4
Duke          4
George        4
Louis         4
             ..
Keet          1
Dixie         1
Ulysses       1
Furry         1
Bearded       1
Sundance      1
Helvetica     1
Be            1
Hero          1
Aldrick       1
Naphaniel     1
Halo          1
Satan         1
Mairi         1
Jubilant      1
Koko          1
Petrick       1
Strider       1
Glacier       1
River         1
Darla         1
Cindy         1
Charleson     1
Banjo         1
Howie         1
Samsom        1
Dew           1
Raphael       1
Rumble        1
Rooney        1
Name: name, Length: 942,

In [83]:
df_archive2.head()

tweet_id           timestamp  \
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   
5  891087950875897856 2017-07-29 00:08:17   

                                                                                                                                                                 text  \
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
5  Here we have a majestic great white breaching off South Africa's coast. Absolutely h*ckin breathtaking. 13/10 (IG: tucker_marlo) #BarkWeek https://t.co/kQ04fDDRmh   

                                                                                                                     expanded_urls  \
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   
5                                                                  https://twitter.com/dog_rates/status/891087950875897856/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper  ...    \
1              13.0                10.0     Tilly  None    None   None  ...     
2              12.0                10.0    Archie  None    None   None  ...     
3              13.0                10.0     Darla  None    None   None  ...     
4              12.0                10.0  Franklin  None    None   None  ...     
5              13.0                10.0       NaN  None    None   None  ...     

  img_num                        p1   p1_conf p1_dog                  p2  \
1     1.0                 Chihuahua  0.323581   True            Pekinese   
2     1.0                 Chihuahua  0.716012   True            malamute   
3     1.0               paper_towel  0.170278  False  Labrador_retriever   
4     2.0                    basset  0.555712   True    English_springer   
5     1.0  Chesapeake_Bay_retriever  0.425595   True       Irish_terrier   

    p2_conf p2_dog                           p3   p3_conf p3_dog  
1  0.090647   True                     papillon  0.068957   True  
2  0.078253   True                       kelpie  0.031379   True  
3  0.168086   True                      spatula  0.040836  False  
4  0.225770   True  German_short-haired_pointer  0.175219   True  
5  0.116317   True              Indian_elephant  0.076902  False  

[5 rows x 22 columns]

In [84]:
df_archive2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1674 entries, 1 to 2116
Data columns (total 22 columns):
tweet_id              1674 non-null object
timestamp             1674 non-null datetime64[ns]
text                  1674 non-null object
expanded_urls         1674 non-null object
rating_numerator      1674 non-null float64
rating_denominator    1674 non-null float64
name                  1321 non-null object
doggo                 1674 non-null object
floofer               1674 non-null object
pupper                1674 non-null object
puppo                 1674 non-null object
jpg_url               1674 non-null object
img_num               1674 non-null float64
p1                    1674 non-null object
p1_conf               1674 non-null float64
p1_dog                1674 non-null object
p2                    1674 non-null object
p2_conf               1674 non-null float64
p2_dog                1674 non-null object
p3                    1674 non-null object
p3_conf             

### Tidiness
1. Melt 'doggo' 'floofer' 'pupper' 'puppo' in `df_archive`.

In [85]:
df_archive2.head(1)

tweet_id           timestamp  \
1  892177421306343426 2017-08-01 00:17:27   

                                                                                                                                                                 text  \
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

                                                     expanded_urls  \
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator   name doggo floofer pupper  ...    \
1              13.0                10.0  Tilly  None    None   None  ...     

  img_num         p1   p1_conf p1_dog        p2   p2_conf p2_dog        p3  \
1     1.0  Chihuahua  0.323581   True  Pekinese  0.090647   True  papillon   

    p3_conf p3_dog  
1  0.068957   True  

[1 rows x 22 columns]

In [86]:
df_archive2['stage'] = df_archive2.doggo + df_archive2.floofer + df_archive2.pupper + df_archive2.puppo
df_archive2.head(1)

tweet_id           timestamp  \
1  892177421306343426 2017-08-01 00:17:27   

                                                                                                                                                                 text  \
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

                                                     expanded_urls  \
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator   name doggo floofer pupper  \
1              13.0                10.0  Tilly  None    None   None   

         ...                p1   p1_conf  p1_dog        p2   p2_conf p2_dog  \
1        ...         Chihuahua  0.323581    True  Pekinese  0.090647   True   

         p3   p3_conf p3_dog             stage  
1  papillon  0.068957   True  NoneNoneNoneNone  

[1 rows x 23 columns]

In [87]:
df_archive2.stage.value_counts()

NoneNoneNoneNone        1414
NoneNonepupperNone       168
doggoNoneNoneNone         54
NoneNoneNonepuppo         21
doggoNonepupperNone        8
NoneflooferNoneNone        7
doggoNoneNonepuppo         1
doggoflooferNoneNone       1
Name: stage, dtype: int64

In [88]:
df_archive2.loc[df_archive2.stage == 'NoneNoneNoneNone', 'stage'] = 'None'
df_archive2.loc[df_archive2.stage == 'NoneNonepupperNone', 'stage'] = 'pupper'
df_archive2.loc[df_archive2.stage == 'doggoNoneNoneNone', 'stage'] = 'doggo'
df_archive2.loc[df_archive2.stage == 'NoneNoneNonepuppo', 'stage'] = 'puppo'
df_archive2.loc[df_archive2.stage == 'doggoNonepupperNone', 'stage'] = 'multiple'
df_archive2.loc[df_archive2.stage == 'NoneflooferNoneNone', 'stage'] = 'floofer'
df_archive2.loc[df_archive2.stage == 'doggoflooferNoneNone', 'stage'] = 'multiple'
df_archive2.loc[df_archive2.stage == 'doggoNoneNonepuppo', 'stage'] = 'multiple'

In [89]:
df_archive2.stage.value_counts()

None        1414
pupper       168
doggo         54
puppo         21
multiple      10
floofer        7
Name: stage, dtype: int64

In [90]:
# drop individual stages columns
df_archive2.drop(['floofer','doggo','pupper', 'puppo'], axis=1, inplace = True)
df_archive2.head(1)

tweet_id           timestamp  \
1  892177421306343426 2017-08-01 00:17:27   

                                                                                                                                                                 text  \
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

                                                     expanded_urls  \
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator   name  \
1              13.0                10.0  Tilly   

                                           jpg_url  img_num         p1  \
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg      1.0  Chihuahua   

    p1_conf p1_dog        p2   p2_conf p2_dog        p3   p3_conf p3_dog stage  
1  0.323581   True  Pekinese  0.090647   True  papillon  0.068957   True  None

In [91]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


2. Join all data frames into `twitter_archive_master.csv`.

In [92]:
twitter_archive_master = df_archive2.merge(df_json, on=['tweet_id'], how='left')

In [93]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1674 entries, 0 to 1673
Data columns (total 21 columns):
tweet_id              1674 non-null object
timestamp             1674 non-null datetime64[ns]
text                  1674 non-null object
expanded_urls         1674 non-null object
rating_numerator      1674 non-null float64
rating_denominator    1674 non-null float64
name                  1321 non-null object
jpg_url               1674 non-null object
img_num               1674 non-null float64
p1                    1674 non-null object
p1_conf               1674 non-null float64
p1_dog                1674 non-null object
p2                    1674 non-null object
p2_conf               1674 non-null float64
p2_dog                1674 non-null object
p3                    1674 non-null object
p3_conf               1674 non-null float64
p3_dog                1674 non-null object
stage                 1674 non-null object
retweet_count         1674 non-null int64
favorite_count      

In [94]:
twitter_archive_master.head(1)

tweet_id           timestamp  \
0  892177421306343426 2017-08-01 00:17:27   

                                                                                                                                                                 text  \
0  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator   name  \
0              13.0                10.0  Tilly   

                                           jpg_url  img_num         p1  \
0  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg      1.0  Chihuahua   

        ...        p1_dog        p2   p2_conf  p2_dog        p3   p3_conf  \
0       ...          True  Pekinese  0.090647    True  papillon  0.068957   

   p3_dog stage retweet_count  favorite_count  
0    True  None          6514           33819  

[1 rows x 21 columns]

In [95]:
# save to csv
twitter_archive_master.to_csv('twitter_archive_master.csv', index=False)